src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java

```diff
@@ -218,30 +223,46 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
 
         private final Duration pollTimeout = receiverOptions.pollTimeout();
 
+        private final AtomicBoolean pausedByUs = new AtomicBoolean();
+
+        private final Set<TopicPartition> pausedByUser = new HashSet<>();
+
+        private final AtomicBoolean scheduled = new AtomicBoolean();
+
         @Override
         public void run() {
             try {
+                this.scheduled.set(false);
                 if (isActive.get()) {
                     // Ensure that commits are not queued behind polls since number of poll events is
                     // chosen by reactor.
                     commitEvent.runIfRequired(false);
                     long r = requested;
                     if (r > 0) {
                         if (!awaitingTransaction.get()) {
-                            consumer.resume(consumer.assignment());
+                            if (pausedByUs.getAndSet(false)) {
+                                Set<TopicPartition> toResume = new HashSet<>(consumer.assignment());
+                                toResume.removeAll(this.pausedByUser);
+                                this.pausedByUser.clear();
+                                consumer.resume(toResume);
+                                log.debug("Resumed");
+                            }
                         } else {
-                            consumer.pause(consumer.assignment());
-                            schedule();
+                            if (!pausedByUs.getAndSet(true)) {
+                                this.pausedByUser.addAll(consumer.paused());
+                                consumer.pause(consumer.assignment());
+                                log.debug("Paused - awaiting transaction");
+                            }
                         }
-                    } else {
+                    } else if (!pausedByUs.getAndSet(true)) {
+                        this.pausedByUser.addAll(consumer.paused());
                         consumer.pause(consumer.assignment());
+                        log.debug("Paused - back pressure");
                     }
 
                     ConsumerRecords<K, V> records = consumer.poll(pollTimeout);
                     if (isActive.get()) {
-                        if (r > 1 || commitEvent.inProgress.get() > 0) {
-                            schedule();
-                        }
+                        schedule();
                     }
 
                     Operators.produced(REQUESTED, ConsumerEventLoop.this, 1);
```

Comment:
I had that before, but restored it because `onRequest()` was never called (it's called downstream from the sink) so we never resumed the consumer.

That might have been on an earlier iteration of this patch; I'll look again tomorrow.

Reply:
See https://github.com/reactor/reactor-kafka/issues/183#issuecomment-796229889

Reply:
After further review, I had to pull that change because `KafkaReceiverTests.userPause()` now fails, because now there is no thread to call `onRequest()` to resume the consumer. Need to give this some more thought.

Reply:
I was mistaken; that code looks fine; there is something else going on that I need to track down tomorrow.

Reply:
Thanks